### PR TITLE
D8MT-447

### PR DESCRIPTION
**Pull request by samuelsolis**
### Description
Added the **SetWorkflowState** process plugin inside a new submodule (oe_migration_set_workflow_state) that allows a better mapping between the D7 workflow states and the new one.

It's a submodule because it has a dependency of the _workflow_ module.

### Change log

- Added:
modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
modules/oe_migration_workflow/tests/Plugin/migrate/process/SetWorkflowStateTest.php
modules/oe_migration_workflowoe_migration_workflow.info.yml

- Changed:
README.md

### CODE REVIEW

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (upchuk) [inline modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: You can remove the @throws

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (upchuk) [inline modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: We typically keep these in 1 line, the arguments..

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (upchuk) [inline modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: Problem here is that you are defaulting to a workflow which doesn't exist. It's defined in another component but you are not depending on it. That's not really good. Perhaps it's better to make these value mandatory for each migration. I see you are validating for the existence of the actual workflow but as a standalone component, `'workflows.workflow.oe_corporate_workflow'` means nothing..

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (upchuk) [inline modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: You can remove the @throws (everywhere)

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (upchuk) [inline modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php:192]: This is way more complicated that it needs to be. First of all, why is the expected configuration the workflow config name (ex. `workflows.workflow.oe_corporate_workflow`) instead of the actual entity ID of it (ex `oe_corporate_workflow`). This way your validation can simply e to load the workflow and see if it exists..

Not sure why you checking the entity type dynamically and stuff, it can only be "workflow".

So a lot of this can be simplified.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (upchuk) [inline modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: Why are you hardcoding the `status` field in here? Make this better configurable. What if I have a migration where that value is under a `published` key. You can move this up to the plugin config and default it to `status` if you want.

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: Missing empty line

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: ?

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: Mmm, is this the plugin ID? Isn't it `oe_migration_set_workflow_state `?

Also, this is not reused. You don't need the variable here, you can directly use the value inside `initializePlugin`

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php:49]: The parent creates already the row here. Should you be overriding it completely?

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: Let's keep the same order for faster readability: name, status

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: Unsure if you need this one here either? Better to just pass it to initializePlugin no? More readable

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: Typo

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: snake_case when inside methods

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: `assertEquals` takes the expected parameters first always.

And always please try to use a string or array representation in the expected, and not a variable. This makes the assertion clear as well as prevents possible false positives due to variable value change. Also, commenting assertions help readability a lot.

```
// The resulting value should be "draft" because that is the default plugin configuration that matches with a status of 0.
$this->assertEquals('draft', $statusDestination);
```

Do this everywhere please.

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: This can be moved inside the `testTransform` which you can name `testTransformDefault`. You are testing the same workflow but with different values but are expecting the same results due to them being invalid (missing) workflow statuses.

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: You are not really testing an alter here. You are testing a different workflow.  Not sure it warrants a new test method either. All these tests you can run one after the other in the same test...

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: typo

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php:20]: Missing a test case for `The entity status and the workflow state status don't match`

**Review by upchuk — CHANGES_REQUESTED** [review]

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: You're totally right. Changed in all places.

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: mm, yes, ok. Merged and changed the name to testTransformDefault

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: Yes, maybe I can put it in the same method. Although I don't like to put several tests in the same test method because it's more difficult to debug and find the broken test when it fails.

BTW you're right that the name is not good. It isn't an alter and it's a new/different workflow. I merged it!

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: fixed

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php:20]: True. Added the test case. I added it in a new test method because it's a different use case and the _expectException_ and _expectExceptionMessage_ calls are difficult to manage if we se it in the same method than other asserts.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: done.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: Yes, changed

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: Done, really the workflow_config_name option already was set as mandatory, I don't know when or why it's configured with a _random_ default value.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: Done

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php:192]: Wow, it's true. It simplifies a lot the code and related test. I performed a huge refinement of the code refactoring this part and allowing only the config as a workflow config (eg. `oe_corporate_workflow`). Less dependencies and same functionality.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: I'm not sure about it. Do we need to make this field configurable? I cannot find an use case in which someone need to manage the entity status with a different field. BTW, I added it as a configuration value and use it with default value as "status".

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: done!

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: haha stupid mistake declaring a dummy class for a test. Removed. (Obviously added automatically by the IDE)

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: Yes, I  usually declare it to have more or less the code of the same way, but in this case we can remove it.

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: Changed

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php:49]: I'm not sure how to change it (or if do it). If I use the defined Row in the parent class (as a mock) I have to change a lot of code and mock different calls in every test.

I'm not sure if it's the best way to do it but I checked other tests in core and it does the same thing, declaring a new Row instance in some places to perform the test when it's necessary.

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: mm... maybe, yes. Changed.

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: ups, a Spanish word :)

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: True. Changed!

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (ioannis-ampatzis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: It's not an inherited exception but introduced by our implementation. By removing it, we are leaving the user with an undeclared exception. Since the override also introduces a new parameter, I would like to propose to remove the `{@inheritdoc}` and provide a full description of the constructor, as some core process plugins (Download, EntityExists, FileCopy, FileProcessBase, MachineName, MenuLinkParent, MigrationLookup and SubProcess) do when they override the constructor. Sorry, my omission.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (ioannis-ampatzis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: Two core process plugins (MigrationLookup and MakeUniqueBase) declare the exception, some others don't. I guess that there is not a global rule about that and it relies on coding styles. I personally prefer to provide as much information as possible in the comments of a method and declare the exceptions :)

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (ioannis-ampatzis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php:192]: This is totally true, but I think that the `getStorage` call should be placed in the static `create` method and probably the `$workflow` variable of the `transfom()` method should become a class variable, since it's going to be shared by every row. As it is now, it's called for every row. Mea culpa, sorry.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (ioannis-ampatzis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: Thank you for this comment, it gives the opportunity to request clear guidelines from you (Core Team). Probably I misunderstood, I thought that inside an OE context we could use default values related to the environment. I specifically asked this question during our meeting with Francesco, Marco and Samuel in February 2020. Could you please give us clear directives? Thank you.

With that in mind,

- Since we're in an OpenEuropa and not in a Drupal.org context, the default values serve best the users inside this context and provide the default OE values, that can be overriden.
- You're absolutely right, it means nothing, you can as well use "test" as a default value. In any case, if one of the three values doesn't exist, an exception will be thrown and the migration will never start.
- We do not create any dependency, it's just a string that will be tested for its validity.
- In the same logic, maybe, we shouldn't use default values for draft or published. It's an assumption that another workflow will define these two states.

Thank you.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (ioannis-ampatzis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: I'm sorry but I'm missing the point of this comment.
The plugin, in its initial state, provides three functionalities
- checks if the input value is a valid state
- if not a valid state, defaults to the entity status
- if valid, verifies that the entity status and the state status are in accordance
By rendering the "status" configurable, we lose two of the functionalities.

Also
- I cannot see which other field could be used as a fallback value to set the workflow state when the provided value is NULL or not a valid one
- we open the "doors" to faulty configuration, e.g. users are able to use fields like `title` or `body/0/value`, which doesn't make sense
- we might have to deal with corrupted data and now we're losing the benefit to verify if entity status and workflow state status are in accordance. if we remove the accordance validation, we might end with nodes that have "status = 1" and "moderation_state = draft"

@samuelsolis , when you refactor code, could you please ensure that the class description, the comments and the code are talking about the same thing.
Comparing a typecasted input value to int, to the state status doesn't make any sense, it looks more like a bug, could you please remove it?

Thanks.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: You're right Ioannis, Maybe it's not an error. I changed it because I didn't know the rule and I just tried to resolve al comments in order to allow the merge. So, please, let me know if I have to change back it. :)

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: Yes, you're right although I'm not sure exactly the meaning of "status" in this contexts. I'm waiting for a Daniel answer , but for sure I can remove this check and check again the comments to ensure that it's correct.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (upchuk) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: No, thats not the point. We do not put @throws in the docblock according to our coding standards. Anywhere. 

RE {@inheritdoc}, you can leave it. Thats not an issue if you don't capture all params.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (upchuk) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: OK, so to explain:

> Probably I misunderstood, I thought that inside an OE context we could use default values related to the environment.

That's not the point. You can use a default value from oe_ context, however, you need to put a dependency on that other component. Otherwise there is no connection. The two are completely decoupled. So if you wanna default to the corporate workflow, you need to composer.json depend on the editorial component. Otherwise, cannot make this assumption. That's the nature of decoupled components.

> Since we're in an OpenEuropa and not in a Drupal.org context, the default values serve best the users inside this context and provide the default OE values, that can be overriden.

Yes, but see my point above.

> In any case, if one of the three values doesn't exist, an exception will be thrown and the migration will never start.

yeah, of course, they need to be set in the migration, that's fine.

> In the same logic, maybe, we shouldn't use default values for draft or published. It's an assumption that another workflow will define these two states.

This is different. 1. It doesn't imply a value provided by another module. And 2, they are values that are gonna be like that in 90% of cases. Just like the "status" field name being "status". We cannot assume it is, but we can assume most are so we default to it. The 10% can change it. But it's different vs the corporate workflow assumption because that value couples it with another component.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (upchuk) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: We do not add @throws in openeuropa components.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (upchuk) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: > I'm not sure about it. Do we need to make this field configurable? I cannot find an use case in which someone need to manage the entity status with a different field. 

You are looking at this purely from the perspective of migrating froma D7 database. This component can be used for migrating from wherever. If I have a CSV file where the "status" field is called published, I should be able to still use it the same way.

> By rendering the "status" configurable, we lose two of the functionalities.

Absolutely not true. The fact that we can specify the name of the field should have no impact in the provided benefits of it. I asked to make only the field *name* configurable. 

> I cannot see which other field could be used as a fallback value to set the workflow state when the provided value is NULL or not a valid one

No other field would be used, but the field NAME can be different. Don't think only at D7 - D8 migrations.

> we open the "doors" to faulty configuration, e.g. users are able to use fields like title or body/0/value, which doesn't make sense

yes, it doesnt make sense. But there aren't simple "users" who create migrations. Whoever uses this component is responsible for writing correct migrations. You, for example, will use the "status" field. What users are you afraid of?

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php:49]: ok

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: @samuelsolis Yes, but you left there the variable.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (upchuk) [inline modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: All @throws must be removed.

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (upchuk) [inline modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: Param misses doc

**Review by upchuk — CHANGES_REQUESTED** [review]

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php]: Done, I misunderstood the first comment about it. I already removed all @thows .

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php:192]: I changed the code for adding the getStorage call in the create method (you're right, it's better) but I'm not sure about the other point (to set the $workflow var as a class property) so I think that for the moment we can continue without change it. BTW obviously we can change it in the future.

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: :facepalm: fixed

File: modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/tests/src/Unit/Plugin/migrate/process/SetWorkflowStateTest.php]: Added.

File: modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php
Comment (samuelsolis) [reply modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php:192]: I have to add that this little change (move the `getStorage` method to the create method) improve and decrease the test complexity removing a (now useless) mock :D

**Review by upchuk — APPROVED** [review]

**Pull request merged**
